### PR TITLE
Add support for Ubuntu Xenial in development

### LINKED ISSUE #3075: Upgrade Ubuntu operating system in development
`ubuntu/trusty64` is out of date, has trouble with SNI validation at times.

### PR DESCRIPTION
**Pull request by hectcastro**
## Overview

- [x] Vagrant base box updated for all virtual machines
- [x] Services responsible for consuming logs and metrics removed from `services`
- [x] `services` virtual machine provisions cleanly
- [x] Services on `services` virtual machine are functional
- [x] `tiler` virtual machine provisions cleanly
- [ ] Services on `tiler` virtual machine are functional
- [x] `app` virtual machine provisions cleanly
- [x] Services on `app` virtual machine are functional
- [x] `worker` virtual machine provisions cleanly
- [x] Services on `worker` virtual machine are functional
- ...

Fixes #3075 

### Demo

Screen Shot 2019-06-12 at 7 10 03 PM

### Notes

- Analysis jobs completed, but almost always returned table cells with 0s
- Delineate watershed did not work because I do not have the necessary data volume
- Tile service is not functional
- CI [hangs](http://civicci01.internal.azavea.com/job/model-my-watershed-pull-requests/4026/console) on one of the Testem tests.

## Testing Instructions

- Destroy all of the local virtual machines and clear out the `azavea.*` roles
- Follow the instructions in the `README` to setup the local development environment

```console
$ vagrant up
$ ./scripts/manage.sh migrate
```

From within the `app` virtual machine:

```console
$ cd /vagrant
./scripts/setupdb.sh -b  # These can take a few minutes
./scripts/setupdb.sh -s
```

- Interact with the application to determine what functionality is not working properly

### CODE REVIEW

File: deployment/ansible/roles/model-my-watershed.app/templates/systemd-app.service.j2 [hunk 1]
```diff
@@ -0,0 +1,26 @@
+[Unit]
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.app/templates/systemd-app.service.j2]: This uses a template from Gunicorn for how to manage Gunicorn processes with `systemd`.

File: deployment/ansible/roles/model-my-watershed.base/tasks/dependencies.yml [hunk 2]
```diff
@@ -1,11 +1,11 @@
 ---
 - name: Install Geospatial libraries
-  apt: pkg={{ item }} state=present
-  with_items:
-    - "binutils=2.24*"
-    - "libproj-dev=4.8.0*"
-    - "gdal-bin=1.10.1*"
-    - "libgdal1-dev=1.10.1*"
+  apt:
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.base/tasks/dependencies.yml]: Ubuntu 16.04 specific versions.

File: deployment/ansible/roles/model-my-watershed.base/tasks/main.yml [hunk 3]
```diff
@@ -1,4 +1,3 @@
 ---
-- { include: chef-and-puppet.yml, when: "['development', 'test'] | some_are_in(group_names)" }
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.base/tasks/main.yml]: These services do not exist on the Bento base images.

File: deployment/ansible/roles/model-my-watershed.celery-worker/templates/systemd-celeryd.service.j2 [hunk 4]
```diff
@@ -0,0 +1,42 @@
+[Unit]
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.celery-worker/templates/systemd-celeryd.service.j2]: Uses a Celery provided template to manage Celery processes with `systemd`.

File: deployment/ansible/roles/model-my-watershed.app/templates/systemd-app.service.j2 [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ExecReload=/bin/kill -s HUP $MAINPID
+ExecStop=/bin/kill -s TERM $MAINPID
+ExecStopPost=/bin/rm -rf /run/mmw-app
+StandardOutput=syslog
+StandardError=syslog
+SyslogIdentifier=gunicorn
+
+[Install]
+{% if ['development', 'test'] | some_are_in(group_names) -%}
+WantedBy=opt_app.mount
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.app/templates/systemd-app.service.j2]: Service should not start until the Vagrant shared folder exists within the context of the virtual machine.

File: deployment/ansible/roles/model-my-watershed.rwd/tasks/app.yml [hunk 6]
```diff
@@ -3,14 +3,23 @@
...
 
-- name: Pull RWD docker container image
-  command: /usr/bin/docker pull {{ rwd_docker_image }}
-
 - name: Configure RWD service definition
-  template: src=upstart-mmw-rwd.conf.j2
-            dest=/etc/init/mmw-rwd.conf
-  notify:
-    - Restart mmw-rwd
+  docker_service:
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.rwd/tasks/app.yml]: Migrated to a Docker managed service vs. wrapping Docker commands in Upstart/`systemd`.

File: deployment/ansible/roles/model-my-watershed.tiler/defaults/main.yml [hunk 7]
```diff
@@ -1,7 +1,3 @@
 ---
 tiler_home: /opt/tiler
 tiler_deploy_branch: "master"
-
-tiler_log: /var/log/mmw-tiler.log
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.tiler/defaults/main.yml]: Log rotation no longer necessary for this service. Logs are routed directly to `syslog` via `systemd`.

File: deployment/ansible/roles/model-my-watershed.tiler/tasks/dependencies.yml [hunk 8]
```diff
@@ -1,11 +1,19 @@
 ---
 - name: Install canvas rendering dependencies
-  apt: pkg={{ item }} state=present
-  with_items:
-    - "libcairo2-dev=1.13.*"
-    - "libpango1.0-dev=1.36.*"
-    - "libjpeg8-dev=8c*"
-    - "libgif-dev=4.1.*"
+  apt:
+    pkg: ["libcairo2-dev=1.14.*",
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.tiler/tasks/dependencies.yml]: Ubuntu 16.04 specific package versions.

File: deployment/ansible/roles/model-my-watershed.tiler/templates/nginx-tiler.conf.j2 [hunk 9]
```diff
@@ -2,7 +2,8 @@ server {
   listen *:80;
   server_name _;
 
-  access_log /var/log/nginx/mmw-tiler.access.log logstash_json;
+  error_log syslog:server=unix:/dev/log,facility=local7,tag=nginx_tiler,severity=error;
```
Comment (hectcastro) [inline deployment/ansible/roles/model-my-watershed.tiler/templates/nginx-tiler.conf.j2]: Write directly to `syslog` from Nginx.

File: deployment/cfn/application.py [hunk 10]
```diff
@@ -17,14 +17,11 @@
 from utils.constants import (
     ALLOW_ALL_CIDR,
     EC2_INSTANCE_TYPES,
-    GRAPHITE,
```
Comment (hectcastro) [inline deployment/cfn/application.py]: All of the changes in this file should really not be in this PR because they are staging specific concerns. I ended up removing them because I was interactively going through find/replace.

Actually testing that these changes are working should be deferred to another PR creation/review cycle.

File: src/mmw/apps/core/mail/backends/boto_ses_mailer.py [hunk 11]
```diff
@@ -13,7 +13,6 @@
 from django.core.mail.backends.base import BaseEmailBackend
 from django.core.exceptions import ImproperlyConfigured
 
-from django_statsd.clients import statsd
```
Comment (hectcastro) [inline src/mmw/apps/core/mail/backends/boto_ses_mailer.py]: Statsite/StatsD has been removed, so application level libraries interacting with those services has been removed as well.

File: src/mmw/apps/core/mail/backends/boto_ses_mailer.py [hunk 12]
```diff
@@ -66,18 +65,16 @@ def _send_message(self, email_message):
...
 
     def _log_message_failure(self, email_message, trace):
-        statsd.incr('email.message.failure')
         logger.error('Exception raised while sending email '
                      'with subject "%s" to %s: %s' % (
                          email_message.subject, email_message.to, trace))
 
     def _log_quota(self, quota):
-        statsd.gauge('email.quota', quota)
+        pass
```
Comment (hectcastro) [inline src/mmw/apps/core/mail/backends/boto_ses_mailer.py]: I think that we should make a decision to replace the implementation here, or remove `_log_quota` altogether. Leaning toward the latter, but wanted to surface it for review.

**Comment by rajadain** [comment]
Provisioning in a fresh check-out failed for the `worker` with the following error:

```
RUNNING HANDLER [azavea.nginx : Restart Nginx] *********************************
Thursday 20 June 2019  16:55:13 -0400 (0:00:00.104)       0:09:35.197 *********
fatal: [worker]: FAILED! => {"changed": false, "msg": "Unable to start service nginx: Job for nginx.service failed because the control process exited with error code. See \"systemctl status nginx.service\" and \"journalctl -xe\" for details.\n"}
```

Details:

```
> vagrant ssh worker -c 'sudo systemctl status nginx.service'
● nginx.service - A high performance web server and a reverse proxy server
   Loaded: loaded (/lib/systemd/system/nginx.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2019-06-20 20:55:14 UTC; 1min 47s ago
     Docs: man:nginx(8)
  Process: 3169 ExecStartPre=/usr/sbin/nginx -t -q -g daemon on; master_process on; (code=exited, status=1/FAILURE)

Jun 20 20:55:14 worker systemd[1]: Starting A high performance web server and a reverse proxy server...
Jun 20 20:55:14 worker nginx[3169]: nginx: [emerg] unknown log format "logstash_json" in /etc/nginx/sites-enabled/geoprocessing:5
Jun 20 20:55:14 worker nginx[3169]: nginx: configuration file /etc/nginx/nginx.conf test failed
Jun 20 20:55:14 worker systemd[1]: nginx.service: Control process exited, code=exited status=1
Jun 20 20:55:14 worker systemd[1]: Failed to start A high performance web server and a reverse proxy server.
Jun 20 20:55:14 worker systemd[1]: nginx.service: Unit entered failed state.
Jun 20 20:55:14 worker systemd[1]: nginx.service: Failed with result 'exit-code'.
Connection to 127.0.0.1 closed.
```

Looking in to `journalctl -xe`, I see:

```
-- Unit nginx.service has begun starting up.
Jun 20 20:55:14 worker nginx[3169]: nginx: [emerg] unknown log format "logstash_json" in /etc/nginx/sites-enabled/geoprocessing:5
Jun 20 20:55:14 worker nginx[3169]: nginx: configuration file /etc/nginx/nginx.conf test failed
Jun 20 20:55:14 worker systemd[1]: nginx.service: Control process exited, code=exited status=1
```

**Comment by rajadain** [comment]
Succeeded on `vagrant reload worker --provision`

**Comment by hectcastro** [comment]
> Provisioning in a fresh check-out failed for the `worker` with the following error:
> 
> ```
> RUNNING HANDLER [azavea.nginx : Restart Nginx] *********************************
> Thursday 20 June 2019  16:55:13 -0400 (0:00:00.104)       0:09:35.197 *********
> fatal: [worker]: FAILED! => {"changed": false, "msg": "Unable to start service nginx: Job for nginx.service failed because the control process exited with error code. See \"systemctl status nginx.service\" and \"journalctl -xe\" for details.\n"}
> ```

I believe this was fixed in 12928c0e836f7a6beda0214c0e46b277a252e5dd.

**Comment by hectcastro** [comment]
TY. I agree with deferring Windshaft things to https://github.com/WikiWatershed/model-my-watershed/pull/3109. I also opened https://github.com/WikiWatershed/model-my-watershed/issues/3110 to deal with the Testem test suite.

I'm going to take a closer look at the service logging aspects of this PR in another pass, then merge. Thanks for the end-to-end testing.
